File: SWEA/D5/7206. 숫자 게임/숫자 게임.java

import java.util.HashMap;
import java.util.Scanner;

public class Solution {
	
	static Scanner sc = new Scanner(System.in);
	static HashMap<String, Integer> mp = new HashMap<>();

	static int split(String n, int dep,  int bit) {
		if(mp.containsKey(n)) return mp.get(n);
		if(n.length() <= 1) return 0;
		if(dep == n.length()-1) {
			if(bit == 0) return 0;
			
			int multi = 1;
			
			int temp = 0;
			for(int i=0;i<n.length()-1;i++) {
				if((bit & (1<<i)) != 0) {
					int num = Integer.parseInt(n.substring(temp, i+1));
					multi *= num;
					temp = i+1;
				}
			}
			multi *= Integer.parseInt(n.substring(temp, n.length()));
			String s = String.valueOf(multi);
			
			int ret = split(s, 0, 0);
			mp.put(s, ret);
			return ret + 1;
		}
		int ret = Math.max(split(n, dep+1, bit), split(n, dep+1, bit|(1<<dep)));;
		return ret;
	}

	
	public static void main(String[] args) {
		int t = sc.nextInt();
		for(int i=1;i<=t;i++) {
			String s = sc.next();
			int ans = split(s, 0, 0);
			System.out.println("#" + i + " " + ans);
		}
	}

}

File: 프로그래머스/3/12987. 숫자 게임/숫자 게임.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> A, vector<int> B) {
    int ret = 0;
    sort(A.begin(), A.end());
    sort(B.begin(), B.end());
    
    int size = B.size();
    int iter_B = 0;
    for(auto k : A){
        while(k >= B[iter_B]){
            iter_B++;
            if(iter_B == size) return ret;
        }
        ret++;
        iter_B++;
        if(iter_B == size) return ret;
    }
    
    return ret;
}